kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/http/HttpJsonRequestProcessor.java

```diff
@@ -0,0 +1,124 @@
+        this.requestModel = requestModel;
+        this.pattern = Pattern.compile(uriPattern);
+        this.method = method;
+    }
+
+    protected static int getInt(int position, List<String> queryStringGroups) {
+        return Integer.parseInt(queryStringGroups.get(position));
+    }
+
+    protected static String getString(int position, List<String> queryStringGroups) {
+        return queryStringGroups.get(position);
+    }
+
+    @Override
+    public boolean acceptRequest(FullHttpRequest request) {
+        if (!request.method().name().equals(method)) {
+            return false;
+        }
+        return detectGroups(request) != null;
+    }
+
+    @Override
+    public CompletableFuture<FullHttpResponse> processRequest(FullHttpRequest request) {
+        List<String> groups = detectGroups(request);
+        try (ByteBufInputStream inputStream = new ByteBufInputStream(request.content())) {
+            K decodeRequest;
+            if (requestModel == Void.class) {
+                decodeRequest = null;
+            } else {
+                decodeRequest = MAPPER.readValue((DataInput) inputStream, requestModel);
+            }
+            CompletableFuture<R> result;
+            try {
+                result = processRequest(decodeRequest, groups, request);
+            } catch (Exception err) {
+                result = FutureUtil.failedFuture(err);
+            }
+            return result.thenApply(resp -> {
+                if (resp == null) {
+                    return buildErrorResponse(NOT_FOUND, "Not found", "text/plain");
+                }
+                if (resp.getClass() == String.class) {
+                    return buildStringResponse(((String) resp), RESPONSE_CONTENT_TYPE);
+                } else {
+                    return buildJsonResponse(resp, RESPONSE_CONTENT_TYPE);
+                }
+            }).exceptionally(err -> {
+                log.error("Error while processing request", err);
+                return buildJsonErrorResponse(err);
+            });
+        } catch (IOException err) {
+            log.error("Cannot decode request", err);
+            return CompletableFuture.completedFuture(buildErrorResponse(HttpResponseStatus.BAD_REQUEST,
+                    "Cannot decode request: " + err.getMessage(), "text/plain"));
+        }
+    }
+
+    private List<String> detectGroups(FullHttpRequest request) {
+        String uri = request.uri();
+        // TODO: here we are discarding the query string part
+        // in the future we will probably have to implement
+        // query string parameters
+        int questionMark = uri.lastIndexOf('?');
+        if (questionMark > 0) {
+            uri = uri.substring(0, questionMark);
+        }
+        if (uri.endsWith("/")) {
+            // compatibility with Confluent Schema registry
+            uri = uri.substring(0, uri.length() - 1);
+        }
+        Matcher matcher = pattern.matcher(uri);
```

Comment:
What's an example of a `pattern`? It seems a strange pattern to use regex to parse uri for whatever the purpose.